File: iosApp/iosApp/Services/CultService/CultService.swift
// Created by web3d3v on 12/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

enum CultServiceError: Error {
    case unableToFetch
    case failedProposalId
    case noResponse
}

typealias CultProposalsResponse = (pending: [CultProposal], closed: [CultProposal])

protocol CultService {
    func fetchProposals(
        handler: @escaping (Result<CultProposalsResponse, Error>) -> Void
    )
}

final class DefaultCultService {
    let walletService: WalletService
    
    init(walletService: WalletService) {
        self.walletService = walletService
    }
}

extension DefaultCultService: CultService {

    func fetchProposals(handler: @escaping (Result<CultProposalsResponse, Error>) -> Void) {
        fetchPending(handler: handler)
    }
}

private extension DefaultCultService {
    
    func fetchPending(handler: @escaping (Result<CultProposalsResponse, Error>) -> Void) {
        fetchProposals(type: "pending") { [weak self] result in
            guard let self = self else { return }
            switch result {
            case let .success(pending):
                self.fetchClosed(pending: pending, handler: handler)
            case .failure:
                self.fetchClosed(pending: [], handler: handler)
            }
        }
    }
    
    func fetchClosed(
        pending: [CultProposal],
        handler: @escaping (Result<CultProposalsResponse, Error>) -> Void
    ) {
        fetchProposals(type: "closed") { result in
            var closed = [CultProposal]()
            switch result {
            case let .success(proposals): closed = proposals
            case .failure: break
            }
            handler(.success((pending: pending, closed: closed)))
        }
    }
    
    func fetchProposals(type: String, handler: @escaping (Result<[CultProposal], Error>) -> Void) {
        guard let url = URL(string: "https://api.cultdao.io/proposals/\(type)") else {
            handler(.failure(CultServiceError.unableToFetch))
            return
        }
        URLSession.shared.dataTask(with: url) { (data, response, error) in
            DispatchQueue.main.async {
                if let error = error { return handler(.failure(error)) }
                guard let data = data else { return handler(.failure(CultServiceError.unableToFetch)) }
                do {
                    let result = try JSONDecoder().decode(CultProposalServiceJSON.self, from: data)
                    handler(.success(self.cultProposals(from: result.proposals)))
                } catch let error {
                    handler(.failure(error))
                }
            }
        }.resume()
    }
    
    func cultProposals(from proposals: [CultProposalJSON]) -> [CultProposal] {
        proposals.compactMap { cultProposal(from: $0) }
    }
    
    func cultProposal(from proposal: CultProposalJSON) -> CultProposal? {
        guard let status = status(from: proposal) else { return nil }
        return .init(
            id: proposal.id,
            title: "#" + proposal.id + " " + proposal.description.projectName,
            approved: approvedVotes(from: proposal),
            rejeceted: rejectedVotes(from: proposal),
            endDate: endDate(from: proposal),
            guardianInfo: guardianInfo(from: proposal.description),
            projectSummary: proposal.description.shortDescription,
            projectDocuments: projectDocuments(from: proposal),
            cultReward: cultReward(from: proposal.description),
            rewardDistributions: rewardDistribution(from: proposal.description),
            wallet: proposal.description.wallet,
            status: status,
            stateName: proposal.stateName
        )
    }
    
    func status(from proposal: CultProposalJSON) -> CultProposal.Status? {
        let currentBlock = walletService.blockNumber(network: Network.ethereum())
        guard proposal.startBlock < currentBlock.toDecimalString() else { return nil }
        if proposal.endBlock > currentBlock.toDecimalString() { return .pending
        } else { return .closed }
    }
    
    func guardianInfo(from project: CultProposalJSON.Description) -> CultProposal.GuardianInfo? {
        guard
            let proposal = project.guardianProposal,
            let discord = project.guardianDiscord,
            let address = project.guardianAddress
        else { return nil }
        return .init(
            proposal: proposal,
            discord: discord,
            address: address
        )
    }
    
    func projectDocuments(from proposal: CultProposalJSON) -> [CultProposal.ProjectDocuments] {
        var documents = [CultProposal.ProjectDocuments]()
        if let fileURL = proposal.description.file.url {
            documents.append(
                .init(
                    name: Localized("cult.proposals.result.liteWhitepaper"),
                    documents: [
                        .link(
                            displayName: proposal.description.file,
                            url: fileURL
                        )
                    ]
                )
            )
        }
        let socialDocs = proposal.description.socialChannel.replacingOccurrences(
            of: "\n", with: " "
        )
        let socialDocsUrls = socialDocs.split(separator: " ").compactMap { String($0).url }
        if !socialDocsUrls.isEmpty {
            documents.append(
                .init(
                    name: Localized("cult.proposals.result.socialDocs"),
                    documents: socialDocsUrls.compactMap {
                        .link(displayName: $0.absoluteString, url: $0)
                    }
                )
            )
        }
        if let linkURL = proposal.description.links.url {
            documents.append(
                .init(
                    name: Localized("cult.proposals.result.audits"),
                    documents: [
                        .link(displayName: linkURL.absoluteString, url: linkURL)
                    ]
                )
            )
        } else if !proposal.description.links.isEmpty {
            documents.append(
                .init(
                    name: Localized("cult.proposals.result.audits"),
                    documents: [
                        .note(proposal.description.links)
                    ]
                )
            )
        }
        return documents
    }
    
    func approvedVotes(from proposal: CultProposalJSON) -> Double {
        (Double(proposal.forVotes) ?? 0) * 0.000000000000000001
    }

    func rejectedVotes(from proposal: CultProposalJSON) -> Double {
        (Double(proposal.againstVotes) ?? 0) * 0.000000000000000001
    }

    func endDate(from proposal: CultProposalJSON) -> Double {
        let genesisEpocOffset = 1460999972
        let epocEndBlock = (Int(proposal.endBlock) ?? 0) * 13
        return Double(genesisEpocOffset + epocEndBlock)
    }
    
    func cultReward(from description: CultProposalJSON.Description) -> String {
        Localized(
            "cult.proposal.parsing.cultRewardAllocation",
            description.range
        )
    }
    
    func rewardDistribution(from description: CultProposalJSON.Description) -> String {
        let perString = Localized("cult.proposal.parsing.rewardDistribution.per")
        let rate = description.rate.replacingOccurrences(of: "%", with: "")
        return rate + "% " + perString + " " + description.time
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/services/cult/CultService.kt
package com.sonsofcrypto.web3walletcore.services.cult

import com.sonsofcrypto.web3lib.utils.KeyValueStore
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.extensions.jsonDecode
import com.sonsofcrypto.web3lib.extensions.jsonEncode
import com.sonsofcrypto.web3lib.extensions.stdJson
import com.sonsofcrypto.web3lib.legacy.BlockNumberToTimestampHelper
import com.sonsofcrypto.web3walletcore.extensions.Localized
import com.sonsofcrypto.web3walletcore.services.cult.CultProposal.GuardianInfo
import com.sonsofcrypto.web3walletcore.services.cult.CultProposal.ProjectDocuments
import com.sonsofcrypto.web3walletcore.services.cult.CultProposal.ProjectDocuments.Document.Link
import com.sonsofcrypto.web3walletcore.services.cult.CultProposal.ProjectDocuments.Document.Note
import com.sonsofcrypto.web3walletcore.services.cult.CultProposal.Status.CLOSED
import com.sonsofcrypto.web3walletcore.services.cult.CultProposal.Status.PENDING
import com.sonsofcrypto.web3walletcore.services.cult.CultProposalJSON.Description
import io.ktor.client.HttpClient
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.plugins.logging.LogLevel
import io.ktor.client.plugins.logging.Logger
import io.ktor.client.plugins.logging.Logging
import io.ktor.client.plugins.logging.SIMPLE
import io.ktor.client.request.get
import io.ktor.client.request.headers
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.URLProtocol
import io.ktor.http.Url
import io.ktor.http.path
import io.ktor.http.withCharset
import io.ktor.serialization.kotlinx.json.json
import io.ktor.utils.io.charsets.Charsets
import kotlinx.serialization.Serializable

data class CultServiceResponse(
    val pending: List<CultProposal>,
    val closed: List<CultProposal>,
)

interface CultService {
    @Throws(Throwable::class)
    suspend fun fetch(): CultServiceResponse
}

class DefaultCultService(
    private val store: KeyValueStore,
): CultService{

    private val client: HttpClient by lazy {
        HttpClient {
            Logging { level = LogLevel.NONE; logger = Logger.SIMPLE }
            install(ContentNegotiation) { json(
                stdJson, ContentType.Application.Json.withCharset(
                    Charsets.UTF_8)) }
        }
    }

    @Throws(Throwable::class)
    override suspend fun fetch(): CultServiceResponse {
        var pending: List<CultProposal>
        var closed: List<CultProposal>
        var err: Throwable? = null
        try {
            pending = fetch("pending").toCultProposals(PENDING)
            storePending(pending)
        } catch (e: Throwable) {
            pending = pending()
            err = e
        }
        try {
            closed = fetch("closed").toCultProposals(CLOSED)
            storeClosed(closed)
        } catch (e: Throwable) {
            closed = closed()
            err = e
        }
        if (pending.isEmpty() && closed.isEmpty() && err != null) { throw err }
        return CultServiceResponse(pending(), closed())
    }

    private fun storePending(proposals: List<CultProposal>) {
        store[pendingKey] = jsonEncode(proposals)
    }
    private fun pending(): List<CultProposal> =
        jsonDecode(store[pendingKey] ?: "[]") ?: emptyList()
    private val pendingKey = "pending"

    private fun storeClosed(proposals: List<CultProposal>) {
        store[closedKey] = jsonEncode(proposals)
    }
    private fun closed(): List<CultProposal> =
        jsonDecode(store[closedKey] ?: "[]") ?: emptyList()
    private val closedKey = "closed"

    @Throws(Throwable::class)
    private suspend fun fetch(proposals: String): List<CultProposalJSON> {
        val body = client.get {
            headers {
                append(HttpHeaders.Accept, "application/json")
                append("X-API-KEY", "")
            }
            url {
                protocol = URLProtocol.HTTPS
                host = "api.cultdao.io"
                path("proposals/$proposals")
                parameters.append("format", "json")
            }
        }.bodyAsText().replace("\"description\":\"\"", "\"description\":{}")
        return jsonDecode<CultProposalServiceJSON>(body)?.data ?: emptyList()
    }
}

private fun List<CultProposalJSON>.toCultProposals(
    status: CultProposal.Status
): List<CultProposal> =
    map { it.toCultProposal(status) }

private fun CultProposalJSON.toCultProposal(status: CultProposal.Status): CultProposal {
    return CultProposal(
        id,
        "#" + id + " " + description.projectName,
        approvedVotes(),
        rejectedVotes(),
        endDate(),
        description.guardianInfo(),
        description.shortDescription ?: "",
        projectDocuments(),
        description.cultReward(),
        description.rewardDistribution(),
        description.wallet ?: "",
        status,
        stateName
    )
}

private fun CultProposalJSON.approvedVotes(): Double =
    forVotes.toDouble() * 0.000000000000000001

private fun CultProposalJSON.rejectedVotes(): Double =
    againstVotes.toDouble() * 0.000000000000000001

private fun CultProposalJSON.endDate(): Double =
    BlockNumberToTimestampHelper.timestamp(endBlock, Network.ethereum())?.toDouble() ?: 0.toDouble()

private fun Description.guardianInfo(): GuardianInfo? {
    val proposal = guardianProposal ?: return null
    val discord = guardianDiscord ?: return null
    val address = guardianAddress ?: return null
    return GuardianInfo(proposal, discord, address)
}

private fun CultProposalJSON.projectDocuments(): List<ProjectDocuments> {
    val documents: MutableList<ProjectDocuments> = mutableListOf()
    documents.add(
        ProjectDocuments(
            Localized("cult.proposals.result.liteWhitepaper"),
            listOf(Link(description.file ?: "", description.file ?: ""))
        )
    )
    val socialDocs = description.socialChannel?.replace("\n", " ") ?: ""
    val socialDocsUrls = socialDocs.split(" ").map { it }
    if (socialDocsUrls.isNotEmpty()) {
        documents.add(
            ProjectDocuments(
                Localized("cult.proposals.result.socialDocs"),
                socialDocsUrls.map { Link(it, it) }
            )
        )
    }
    val url = Url.parse(description.links ?: "")
    if (url != null) {
        documents.add(
            ProjectDocuments(
                Localized("cult.proposals.result.audits"),
                listOf(Link(url.toString(), url.toString()))
            )
        )
    } else {
        documents.add(
            ProjectDocuments(
                Localized("cult.proposals.result.audits"),
                listOf(Note(description.links ?: ""))
            )
        )
    }
    return documents
}

private fun Url.Companion.parse(str: String): Url? {
    return try {
        Url(str)
    } catch (e: Throwable) {
        null
    }
}

private fun Description.cultReward(): String =
    Localized("cult.proposal.parsing.cultRewardAllocation", range)

private fun Description.rewardDistribution(): String {
    val perString = Localized("cult.proposal.parsing.rewardDistribution.per")
    val rate = rate?.replace("%", "") ?: ""
    return "$rate% $perString $time"
}
@Serializable
private data class CultProposalServiceJSON(
    val code: Int,
    val data: List<CultProposalJSON>,
)

@Serializable
private data class CultProposalJSON(
    val id: String,
    val proposer: String,
    val eta: String,
    val startBlock: String,
    val endBlock: String,
    val forVotes: String,
    val againstVotes: String,
    val abstainVotes: String,
    val canceled: Boolean,
    val executed: Boolean,
    val description: Description,
    val state: String, // 1-Active, 2-Cancelled, 3-Defeated, 7-Executed
    val stateName: String,
) {
    @Serializable
    data class Description(
        val projectName: String?,
        val shortDescription: String?,
        val file: String?,
        val socialChannel: String?,
        val links: String?,
        val range: String?,
        val rate: String?,
        val time: String?,
        val wallet: String?,
        val guardianProposal: String?,
        val guardianDiscord: String?,
        val guardianAddress: String?,
    )
}
//
//@Serializer(forClass = BigInt::class)
//private object RangeSerializer : KSerializer<Any> {
//
//    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor(
//        "String", PrimitiveKind.STRING
//    )
//
//    override fun serialize(encoder: Encoder, value: String) {
//        encoder.encodeString("${value.toString()}")
//    }
//
//    override fun deserialize(decoder: Decoder): BigInt {
//        return BigInt.from(decoder.decodeString())
//    }
//}